guidance/_grammar.py

```diff
@@ -243,6 +243,28 @@ def match_byte(self, byte):
     def max_tokens(self):
         return 1000000000000
 
+class Box(Terminal):
+    """Container to hold a value that is resolved at a later time. This is useful for recursive definitions."""
+    __slots__ = "_value"
+
+    def __init__(self):
+        super().__init__(temperature=-1, capture_name=None)
+        self._resolved = False
+        self._value = None
+
+    @property
+    def value(self):
+        if self._resolved:
+            return self._value
+        else:
+            raise ValueError("Box is not yet resolved")
+
+    @value.setter
+    def value(self, value):
+        if self._resolved:
```

Comment:
Just making sure we know we're shooting ourselves in the foot if we try to set the same Box's value more than once. Not strictly necessary.

Reply:
"Not strictly necessary" only applies to now, when you're writing this. In six months' time when you're hunting some obscure bug, this could make it much less obscure :-)